src/Compilers/CSharp/Portable/Syntax/LambdaUtilities.cs

```diff
@@ -85,24 +93,33 @@ internal static SyntaxNode TryGetCorrespondingLambdaBody(SyntaxNode oldBody, Syn
                     return IsReducedSelectOrGroupByClause(selectClause, selectClause.Expression) ? null : selectClause.Expression;
 
                 case SyntaxKind.JoinClause:
-                    var oldJoin = (JoinClauseSyntax)oldLambda;
+                    var oldJoin = (JoinClauseSyntax)(oldBody.Parent);
                     var newJoin = (JoinClauseSyntax)newLambda;
                     Debug.Assert(oldJoin.LeftExpression == oldBody || oldJoin.RightExpression == oldBody);
                     return (oldJoin.LeftExpression == oldBody) ? newJoin.LeftExpression : newJoin.RightExpression;
 
                 case SyntaxKind.GroupClause:
-                    var oldGroup = (GroupClauseSyntax)oldLambda;
+                    var oldGroup = (GroupClauseSyntax)(oldBody.Parent);
                     var newGroup = (GroupClauseSyntax)newLambda;
                     Debug.Assert(oldGroup.GroupExpression == oldBody || oldGroup.ByExpression == oldBody);
                     return (oldGroup.GroupExpression == oldBody) ?
                         (IsReducedSelectOrGroupByClause(newGroup, newGroup.GroupExpression) ? null : newGroup.GroupExpression) : newGroup.ByExpression;
 
                 case SyntaxKind.LocalFunctionStatement:
-                    var newLocalFunction = (LocalFunctionStatementSyntax)newLambda;
-                    return (SyntaxNode)newLocalFunction.Body ?? newLocalFunction.ExpressionBody;
+                    return GetLocalFunctionBody((LocalFunctionStatementSyntax)newLambda);
 
                 default:
-                    throw ExceptionUtilities.UnexpectedValue(oldLambda.Kind());
+                    throw ExceptionUtilities.UnexpectedValue(newLambda.Kind());
+            }
+        }
+
+        public static SyntaxNode GetNestedFunctionBody(SyntaxNode nestedFunction)
+        {
+            switch (nestedFunction)
+            {
+                case AnonymousFunctionExpressionSyntax anonymousFunctionExpressionSyntax: return anonymousFunctionExpressionSyntax.Body;
+                case LocalFunctionStatementSyntax localFunctionStatementSyntax: return (CSharpSyntaxNode)localFunctionStatementSyntax.Body ?? localFunctionStatementSyntax.ExpressionBody.Expression;
```

Comment:
The EnC analysis expects Expressions and Blocks to be bodies of lambdas of all forms. 

This stems from inconsistencies in C# syntax. AnonymousFunctionExpressionSyntax doesn't have ArrowExpression, while LocalFunctions do. Lambdas in queries are also just expressions. We had to define a common "lambda body" abstraction, which is Expression or Block.

---
In reply to: [134613485](https://github.com/dotnet/roslyn/pull/21599#discussion_r134613485) [](ancestors = 134613485)